src/layers/decoder_block.h

```diff
@@ -0,0 +1,348 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================================
+#pragma once
+#include <string>
+#include "decoder_layer.h"
+#include "dtype.h"
+#include "kvcache_mgr.h"
+#include "messenger.h"
+#include "weight_util.h"
+
+template <typename ATTN_CLS, typename MLP_CLS, typename KVCacheT, bool ATTN_MLP_PARALLEL>
+class DecoderBlock {
+public:
+    using DECODER = Decoder<ATTN_CLS, MLP_CLS>;
+
+    DecoderBlock(DecoderContext *ctx, const std::string &modelPath, int layers, xft::DataType dt) {
+        if (layers % ctx->ppSize != 0) {
+            std::cerr << "Warning: layers cannot be evenly divided by pipeline parallel stage size(ppSize)."
+                      << std::endl;
+            std::exit(-1);
+        }
+
+        int layersOnDuty = layers / ctx->ppSize;
+        int startLayer = ctx->ppRank * layersOnDuty;
+        for (int i = startLayer; i < startLayer + layersOnDuty; ++i) {
+            auto pdec = new DECODER(ctx, i);
+            if (dt == xft::DataType::int8) {
+                this->setDecoderWeights<int8_t>(ctx, pdec, modelPath, i);
+            } else if (dt == xft::DataType::int4) {
+                this->setDecoderWeights<uint4x2_t>(ctx, pdec, modelPath, i);
+            } else if (dt == xft::DataType::fp32) {
+                this->setDecoderWeights<float>(ctx, pdec, modelPath, i);
+            } else {
+                std::cerr << "Error: The data type is NOT supported." << std::endl;
+                std::exit(-1);
+            }
+            this->decoders.push_back(pdec);
+        }
+    }
+
+    virtual ~DecoderBlock() {
+        for (auto dec : this->decoders) {
+            delete dec;
+        }
+    }
+
+    // To make it compatible with the old impl.
+    DECODER *get(int layerId) { return this->decoders[layerId]; }
+
+    int size() const { return this->decoders.size(); }
+
+    template <typename InT, typename OutT>
+    void forward(DecoderContext *ctx, std::vector<xft::SequenceMeta *> &seqs, InT *input, OutT *output) {
+        using AttnOutT = typename AttnTypeExtractor<ATTN_CLS>::Tout;
+
+        Messenger &messenger = Messenger::getInstance();
+        xft::KVCacheMgr &kvCacheMgr = xft::KVCacheMgr::instance();
+
+        // Data preparation
+        std::vector<int> seqIDs(seqs.size());
+        size_t totInSeqLen = 0;
+        for (int i = 0; i < seqs.size(); ++i) {
+            seqIDs[i] = seqs[i]->getSequenceID();
+            totInSeqLen += seqs[i]->getInputSeqLen();
+        }
+
+        kvCacheMgr.prepareCache(seqIDs);
```

Comment:
I prefer to prepare KVCache at a high level api.
We can assume all seqs' KVCache is ready.

Reply:
commented and added TODO